File: kotlin/src/LogEvent.kt
// Copyright 2018 Sourcerer Inc. All Rights Reserved.
// Author: Alexander Surkov (alex@sourcerer.io)

package sourcererio.vacuumanalytics

import com.github.kittinunf.fuel.core.FuelError
import com.github.kittinunf.fuel.core.FuelManager
import com.github.kittinunf.fuel.core.Method
import com.github.kittinunf.fuel.core.Request
import com.github.kittinunf.fuel.core.Response
import java.net.HttpCookie
import java.util.Date

/**
 * Elastic LogStash logging.
 */
class LogEvent {
    // Unique identifier of the session.
    private var session: String = ""

    private val fuelManager = FuelManager()

    /**
     * Creates a LogEvent instance.
     *
     * @param path [in] path to send logs to,
     * @param res  [in] HTTP response of successful user authorization,
     * @param sid  [in] session id of the HTTP response.
     */
    constructor(path: String,
                res: Response, sid: String) {
        res.headers["Set-Cookie"]?.
            flatMap { HttpCookie.parse(it) }?.
            find { it.name == sid }?.
            let { session = it.value }

        fuelManager.basePath = path
    }

    /**
     * Logs error.
     */
    fun error(message: String) {
        event("error", message)
    }

    /**
     * Logs event.
     */
    fun event(type: String, message: String) {
        event(type, listOf("message" to message))
    }

    /**
     * Logs event.
     */
    fun event(type: String, fields: List<Pair<String, Any?>>) {
        var params = mutableListOf<Pair<String, Any?>>(
            "session" to session,
            "type" to type,
            "timestamp" to Date().getTime()
        )

        fuelManager.
            request(
                Method.POST,
                "",
                params.plus(fields)
            )
            .response { _, _, result ->
                val (_, err) = result
                if (err != null) {
                    throw Exception("FAILED to send logs: $err")
                }
            }
    }
}

File: js/src/LogEvent.js
// Copyright 2018 Sourcerer Inc. All Rights Reserved.
// Author: Alexander Surkov (alex@sourcerer.io)

(function(exports) {

/**
 * Elastic LogStash logging.
 */
class LogEvent
{
  constructor(url, { headers })
  {
    console.assert(url);
    console.assert(headers);

    this.url = url;
    if (headers['set-cookie']) {
      let value = headers['set-cookie'];
      let parts = value.split(';').filter((v) => !!v.trim());
      let cookiePart = parts[0].split('=');
      this.session = cookiePart[1];
    }
  }

  /**
   * Logs an error.
   */
  error(message)
  {
    return this.event('error', { message });
  }

  /**
   * Logs an event of the given type.
   */
  event(type, fields)
  {
    let data = Object.assign({
      session: this.session,
      type,
      timestamp: Date.now()
    }, fields);

    let thisobj = this;
    return new Promise((resolve, reject) => {
      ('XMLHttpRequest' in exports) ?
        thisobj.browserRequest(data, resolve, reject) :
        thisobj.nodeRequest(data, resolve, reject);
    });
  }

  nodeRequest(data, resolve, reject)
  {
    const { URL } = require('url');

    let url = new URL(this.url);
    let protocol = url.protocol.replace(':', '');
    if (protocol != 'http' && protocol != 'https') {
      reject({ error: 'Unsupported protocol' });
      return;
    }

    const options = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      checkServerIdentity: function (host, cert) {
        return undefined;
      }
    };

    const req = require(protocol).request(options, (res) => {
      if (res.statusCode != 200) {
        reject({
          error: `HTTP error`,
          status: res.statusCode
        });
        return;
      }

      let stream = new require('stream').Transform();
      res.on('data', (chunk) => {
        stream.push(chunk);
      });
      res.on('end', () => {
        resolve(new Buffer(stream.read()));
      });
    });
    req.on('error', (e) => {
      reject({
       error: `HTTP error`,
       status: e
     });
    });
    req.end(JSON.stringify(data));
  }

  browserRequest(data, resolve, reject)
  {
    const req = new XMLHttpRequest();
    req.open('POST', this.url, true);

    // Send the proper header information along with the request.
    req.setRequestHeader('Content-type', 'application/json');
    req.withCredentials = true;
    req.onreadystatechange = () => {
      if (req.readyState == 4 && req.status == 200) {
        resolve(req.responseText);
      }
      else {
        reject({
          error: `HTTP error`,
          status: req.status
        });
      }
    }
    req.send(JSON.stringify(data));
  }
};

exports.LogEvent = LogEvent;

})(typeof exports === 'undefined'? this : exports);
